### PR TITLE
buildah-containers: add validation to 'format'

### PR DESCRIPTION
**Pull request by q384566678**
Avoid the following situations：

```
➜  buildah git:(containers-format) sudo buildah containers --format zz
zz
zz
zz
zz
zz
zz
zz
zz
zz
```

After change:
```
➜  buildah git:(containers-format) sudo ./buildah containers --format zz
error format: zz
```

Signed-off-by: Zhou Hao <zhouhao@cn.fujitsu.com>

### CODE REVIEW

File: cmd/buildah/containers.go [hunk 1]
```diff
@@ -239,6 +239,10 @@ func outputContainers(store storage.Store, opts containerOptions, params *contai
 }
 
 func containerOutputUsingTemplate(format string, params containerOutputParams) error {
+	if !strings.HasPrefix(format, "{{.") || !strings.HasSuffix(format, "}}") {
+		return errors.Errorf("error format: %s", format)
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/containers.go]: Can we change the error to ("error invalid format provided: %s", format)

Otherwise LGTM, thanks for the touchup!

Comment (ripcurld) [reply cmd/buildah/containers.go]: I'm afraid if someone will set `--format` to `"container ID: {{.Id}}"` this will return error although it's legit.

Comment (rhatdan) [reply cmd/buildah/containers.go]: I agree. if we want to do the format check to make sure their is json, it should be done with a regex expression.

Comment (q384566678) [reply cmd/buildah/containers.go]: updated, PTAL.

File: cmd/buildah/containers.go [hunk 2]
```diff
@@ -239,6 +240,13 @@ func outputContainers(store storage.Store, opts containerOptions, params *contai
 }
 
 func containerOutputUsingTemplate(format string, params containerOutputParams) error {
+	matched, err := regexp.MatchString("{{.*}}", format)
```
Comment (rhatdan) [inline cmd/buildah/containers.go]: This still does not match the format suggested before.  Perhaps something like
```
matched, err := regexp.MatchString(".*{{.*}}", format)
```
But add some tests of valid formats and invalid formats.

File: cmd/buildah/containers.go [hunk 3]
```diff
@@ -239,6 +240,13 @@ func outputContainers(store storage.Store, opts containerOptions, params *contai
...
+	if err != nil {
+		return err
```
Comment (rhatdan) [inline cmd/buildah/containers.go]: return errors.Wrapf(err, "error validating format provided %s", format)

**Comment by rhatdan** [comment]
Could you add tests including validating that

--format  "container ID: {{.Id}}"

Works, and add this example to the man page.

File: cmd/buildah/containers.go (hunk 2 above)
Comment (q384566678) [reply cmd/buildah/containers.go]: Don't need change. It works.

```
➜  buildah git:(containers-format) sudo ./buildah containers --format "containerID: {{.ContainerID}}"
containerID: e477836657bba9fd76b80fd71c13318dc18a78711102a32b110e72d37187602e
containerID: 09d1845b52485ce3bc431ff8e57b1e4910723d96af6223bf3a421ed6b48a5d6d
containerID: b025ebe4b1d94887e876f2ebc7e410396f5fe4780ec1f80437d36c39cc3189d8
containerID: 5666da5d29003faecbaf44c951ee465402aa2c500e4659b60767ebb7c94b087e
```

**Comment by q384566678** [comment]
>Could you add tests including validating that
--format "container ID: {{.Id}}"

Do you mean this?

sudo ./buildah containers --format "containerID: {{.ContainerID}}"

File: cmd/buildah/containers.go (hunk 2 above)
Comment (rhatdan) [reply cmd/buildah/containers.go]: Excellent.

**Comment by q384566678** [comment]
updated, PTAL.

File: cmd/buildah/containers_test.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		ContainerName: "test-container",
+	}
+
+	formatString := "ContainerID"
+
+	err := containerOutputUsingTemplate(formatString, params)
+	if err == nil || err.Error() != "error invalid format provided: ContainerID" {
+		t.Fatalf("expected error invalid format")
+	}
+}
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/containers_test.go]: I'm not sure these tests will get run automatically with Travis.  They look great, but is there anyway you could change them into  bats format in our /tests directory?

File: cmd/buildah/containers.go [hunk 5]
```diff
@@ -239,6 +240,13 @@ func outputContainers(store storage.Store, opts containerOptions, params *contai
 }
 
 func containerOutputUsingTemplate(format string, params containerOutputParams) error {
+	matched, err := regexp.MatchString("{{.*}}", format)
+	if err != nil {
+		return errors.Wrapf(err, "error validating format provided %s", format)
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/containers.go]: nit of a nit, can you add a colon ':' after provided please?

File: cmd/buildah/containers.go (hunk 2 above)
Comment (TomSweeneyRedHat) [reply cmd/buildah/containers.go]: Is this new code OK with this:

sudo ./buildah containers --format "containerID: {{}}"

i.e. does it hit the error that you have at line 247 or does it do something else?

File: cmd/buildah/containers_test.go (hunk 4 above)
Comment (q384566678) [reply cmd/buildah/containers_test.go]: > could change them into bats format in our /tests directory?

I also wanted to do it at the beginning, but I was not familiar with bats and did not find a suitable method. So only use this method.
I will learn more about bats and hope to find a suitable solution.

**Comment by q384566678** [comment]
> i.e. does it hit the error that you have at line 247 or does it do something else?

No. It will failed. But it is not because of the change.

```
➜  buildah git:(containers-format) sudo ./buildah containers --format "contaienrID: {{}}"
Template parsing error: template: container:1: missing value for command
```

**Comment by TomSweeneyRedHat** [comment]
Thanks for checking on:

```
 buildah git:(containers-format) sudo ./buildah containers --format "contaienrID: {{}}"
Template parsing error: template: container:1: missing value for command
```
That's a valid error in that situation. I was concerned it might go completely off the rails in that instance and not give enough info to the user to figure out what the issue was.

**Comment by rhatdan** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
LGTM

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Comment by rhatdan** [comment]
@rh-atomic-bot r+
@q384566678 Thanks for the PR.

**Pull request closed without merging**